identity-account/src/account/account.rs

```diff
@@ -255,166 +255,152 @@ impl Account {
     new_state: &IdentityState,
     document: &mut IotaDocument,
   ) -> Result<()> {
-    if new_state.integration_generation() == Generation::new() {
-      let method: &TinyMethod = new_state.capability_invocation()?;
-      let location: &KeyLocation = method.location();
+    if self.chain_state().is_new_identity() {
+      let method: &IotaVerificationMethod = new_state.document().default_signing_method()?;
+      let location: KeyLocation = new_state.method_location(
+        method.key_type(),
+        method
+          .id()
+          .fragment()
+          .expect("verification method did not have a fragment")
+          .to_owned(),
+      )?;
 
       // Sign the DID Document with the current capability invocation method
       new_state
-        .sign_data(self.did(), self.storage(), location, document)
+        .sign_data(self.did(), self.storage(), &location, document)
         .await?;
     } else {
-      let method: &TinyMethod = old_state.capability_invocation()?;
-      let location: &KeyLocation = method.location();
+      let method: &IotaVerificationMethod = old_state.document().default_signing_method()?;
+      let location: KeyLocation = new_state.method_location(
+        method.key_type(),
+        method
+          .id()
+          .fragment()
+          .expect("verification method did not have a fragment")
+          .to_owned(),
+      )?;
```

Comment:
Bubbling up comes at little cost, but I'd have to add that variant to the error enum first, which adds to the number of errors a developer has to consider. It pollutes the error enum. No one would actually expect this error when calling `publish_changes`, because the user isn't providing any fragment here. If they had provided an erroneous fragment at some earlier point, they would've expected that method to error.

There seems to be a divide in the Rust community in general, where people say "panicking when encountering a broken invariant is fine", while others avoid it at all cost. I lean towards the former statement, but I'm also okay with us setting the guideline as "avoid panicking" for identity.rs. In that case, I would like us to consider a `FatalError` catch-all variant for these errors, so we bubble them up and give the user a chance to do some clean up or emergency handling. But returning `FatalError("method missing fragment")` seems totally fine here. In fact, Oliver is working on something like that in #487, so we can also take the discussion there, once it's ready.

For now, I'll add the variant and a `TODO` so we can reconsider in a future error-focused PR.

Reply:
>No one would actually expect this error when calling publish_changes, because the user isn't providing any fragment here. If they had provided an erroneous fragment at some earlier point, they would've expected that method to error.
When this is the case it is indeed dangerous to bundle `MethodMissingFragment` together with failures we are able to explain as it is then not clearly conveyed that the program has now entered an unexpected state. 

In general there is a lot to be said about panicking early. It can prevent a corrupted program from producing disastrous side effects and it can also help us catch bugs early on in cases where our assumptions and/or program logic is wrong.  We should also keep in mind that panics in Rust is actually quite graceful in the sense that all `Drop` implementations get to run and it doesn't necessarily affect other threads, not to mention the generally helpful error message it yields. 

Having said that, one can of course question whether our library gets to decide when to panic. An alternative would be to follow the `sled` post mentioned in [this discussion](https://github.com/iotaledger/identity.rs/discussions/460) and return something like `Result<Result<(),ExpectedError>,FatalError>` which the caller may choose to either bubble up or panic from.  

Regardless of what is said above any function that can panic must document this in its documentation, and since one often does not bother documenting private functions, they probably should return the nested `Result` with an outer `FatalError`.